datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+) -> Result<Arc<dyn crate::physical_plan::ExecutionPlan>> {
+    let plan_any = plan.as_any();
+    if let Some(HashJoinExec {
+        left,
+        right,
+        on,
+        filter,
+        join_type,
+        mode,
+        null_equals_null,
+        ..
+    }) = plan_any.downcast_ref::<HashJoinExec>()
+    {
+        match mode {
+            PartitionMode::Partitioned => {
+                let join_key_pairs = extract_join_keys(on);
+                if let Some((
+                    JoinKeyPairs {
+                        left_keys,
+                        right_keys,
+                    },
+                    new_positions,
+                )) = try_reorder(
+                    join_key_pairs.clone(),
+                    parent_required,
+                    &plan.equivalence_properties(),
+                ) {
+                    let new_join_on = if !new_positions.is_empty() {
+                        new_join_conditions(&left_keys, &right_keys)
+                    } else {
+                        on.clone()
+                    };
+                    let new_left =
+                        adjust_input_keys_down_recursively(left.clone(), left_keys)?;
+                    let new_right =
+                        adjust_input_keys_down_recursively(right.clone(), right_keys)?;
+                    Ok(Arc::new(HashJoinExec::try_new(
+                        new_left,
+                        new_right,
+                        new_join_on,
+                        filter.clone(),
+                        join_type,
+                        PartitionMode::Partitioned,
+                        null_equals_null,
+                    )?))
+                } else {
+                    let new_left = adjust_input_keys_down_recursively(
+                        left.clone(),
+                        join_key_pairs.left_keys,
+                    )?;
+                    let new_right = adjust_input_keys_down_recursively(
+                        right.clone(),
+                        join_key_pairs.right_keys,
+                    )?;
+                    Ok(Arc::new(HashJoinExec::try_new(
+                        new_left,
+                        new_right,
+                        on.clone(),
+                        filter.clone(),
+                        join_type,
+                        PartitionMode::Partitioned,
+                        null_equals_null,
+                    )?))
+                }
+            }
+            PartitionMode::CollectLeft => {
+                let new_left = adjust_input_keys_down_recursively(left.clone(), vec![])?;
+                let new_right = match join_type {
+                    JoinType::Inner | JoinType::Right => try_push_required_to_right(
+                        parent_required,
+                        right.clone(),
+                        left.schema().fields().len(),
+                    )?,
+                    JoinType::RightSemi | JoinType::RightAnti => {
```

Comment:
For `RightSemi` and `RightAnti` join, the current schema should be the same with the right one. Therefore, no need to do schema adjustment.